arangod/RocksDBEngine/RocksDBVectorIndex.cpp

```diff
@@ -0,0 +1,482 @@
+bool RocksDBVectorIndex::matchesDefinition(VPackSlice const& /*info*/) const {
+  return false;
+}
+
+void RocksDBVectorIndex::toVelocyPack(
+    arangodb::velocypack::Builder& builder,
+    std::underlying_type<Index::Serialize>::type flags) const {
+  VPackObjectBuilder objectBuilder(&builder);
+  RocksDBIndex::toVelocyPack(builder, flags);
+  builder.add(VPackValue("params"));
+  velocypack::serialize(builder, _definition);
+
+  if (_trainedData && Index::hasFlag(flags, Index::Serialize::Internals)) {
+    builder.add(VPackValue("trainedData"));
+    velocypack::serialize(builder, *_trainedData);
+  }
+}
+
+std::pair<std::vector<VectorIndexLabelId>, std::vector<float>>
+RocksDBVectorIndex::readBatch(std::vector<float>& inputs,
+                              RocksDBMethods* rocksDBMethods,
+                              transaction::Methods* trx,
+                              std::shared_ptr<LogicalCollection> collection,
+                              std::size_t count, std::size_t topK) {
+  TRI_ASSERT(topK * count == (inputs.size() / _definition.dimension) * topK);
+
+  auto flatIndex = createFaissIndex(_quantizer, _definition);
+  RocksDBInvertedLists ril(this, collection.get(), trx, rocksDBMethods, _cf,
+                           _definition.nLists, flatIndex.code_size);
+  flatIndex.replace_invlists(&ril);
+
+  std::vector<float> distances(topK * count);
+  std::vector<faiss::idx_t> labels(topK * count);
+
+  if (_definition.metric == SimilarityMetric::kCosine) {
+    faiss::fvec_renorm_L2(_definition.dimension, count, inputs.data());
+  }
+  flatIndex.search(count, inputs.data(), topK, distances.data(), labels.data(),
+                   nullptr);
+  // faiss returns squared distances for L2, square them so they are returned in
+  // normal form
+  if (_definition.metric == SimilarityMetric::kL2) {
+    std::ranges::transform(distances, distances.begin(),
+                           [](auto const& elem) { return std::sqrt(elem); });
+  }
+
+  return {std::move(labels), std::move(distances)};
+}
+
+// TODO Move away, also remove from MDIIndex.cpp
+auto accessDocumentPath(VPackSlice doc,
+                        std::vector<basics::AttributeName> const& path)
+    -> VPackSlice {
+  for (auto&& attrib : path) {
+    TRI_ASSERT(attrib.shouldExpand == false);
+    if (!doc.isObject()) {
+      return VPackSlice::noneSlice();
+    }
+
+    doc = doc.get(attrib.name);
+  }
+
+  return doc;
+}
+
+/// @brief inserts a document into the index
+Result RocksDBVectorIndex::insert(transaction::Methods& /*trx*/,
+                                  RocksDBMethods* methods,
+                                  LocalDocumentId documentId,
+                                  velocypack::Slice doc,
+                                  OperationOptions const& /*options*/,
+                                  bool /*performChecks*/) {
+  TRI_ASSERT(_fields.size() == 1);
+  auto flatIndex = createFaissIndex(_quantizer, _definition);
```

Comment:
Not very expensive, the quantizer data is, that's why the index holds only a pointer to it. I wish I could avoid this but the reason we do this way is because it is impossible with the faiss design to pass the RocksDB methods to `InvertedLists` otherwise.